### PR TITLE
TLS over an existing connection

### PR DESCRIPTION
**Pull request by art-w**
With @MisterDA , we have a use-case which requires TLS tunneling over an existing TLS connection. This should be pretty easy, but the current `tls-lwt` API only allow TLS negotiation over an `Lwt_unix.file_descr` which limits our ability to use an existing TLS connection as the carrier. This PR adds a new `Tls_lwt.Unix.client_of_channels` function to establish the TLS over input/output `Lwt_io.channel` instead, which were potentially created by a previous TLS.

To explain why TLS-over-TLS is useful, I've attached an example to show how HTTPS proxying work:
- The client first does an HTTPS request to the proxy, and ask it to `CONNECT to_some_server.net:443`
- The proxy hopefully responds with a `200 OK` success code after connecting to the server
- From there, the TLS connection with the proxy remains open, but further messages on that socket are transparently relayed by the proxy between the client and server
- To send an HTTPS request to the now-reachable target server, the client must do a new TLS negotiation with that server (over the existing TLS connection with the proxy)

(Note that this mode of proxying is rather nice as it guarantees that the proxy can't do a man-in-the-middle, since the client is in charge of checking that the nested TLS connection has been established with the expected server certificates!)

I would like to rebase this PR on top of https://github.com/mirleft/ocaml-tls/pull/497 to get rid of the cstructs, but in the meantime, let me know if there's anything else that should be changed! :)

### CODE REVIEW

File: lwt/tls_lwt.ml [hunk 1]
```diff
@@ -31,10 +31,43 @@ module Lwt_cs = struct
     | cs -> write fd cs >>= o (write_full fd) (Cstruct.shift cs)
 end
 
+module Lwt_fd = struct
+  type t = {
+    read  : Cstruct.t -> int Lwt.t ;
+    write : Cstruct.t -> unit Lwt.t ;
+    close : unit -> unit Lwt.t ;
+  }
```
Comment (art-w) [inline lwt/tls_lwt.ml:39]: I didn't want to mess with the carefully crafted `write`/`close` implementation present for file descriptors, but perhaps using a pair of `input/output Lwt_io.channel`s would work instead of a new abstraction? (I'm not sure of the impact, because the doc for lwt channels mentions that they do buffering)

File: lwt/examples/tls_over_tls.ml [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+$ openssl req -x509 -nodes -new -sha256 -days 1024 -newkey rsa:2048 -keyout RootCA.key -out RootCA.pem -subj "/C=US/CN=Example-Root-CA"
+$ openssl x509 -outform pem -in RootCA.pem -out RootCA.crt
+$ cat > domains.ext
+authorityKeyIdentifier=keyid,issuer
+basicConstraints=CA:FALSE
+keyUsage = digitalSignature, nonRepudiation, keyEncipherment, dataEncipherment
+subjectAltName = @alt_names
+[alt_names]
+DNS.1 = localhost
```
[the remark this answers is not in the document]
Comment (art-w) [reply lwt/examples/tls_over_tls.ml:18]: Thanks!

**Comment by hannesm** [comment]
so what is the story of the other backends?

but more importantly, why only a client over an existing connection? wouldn't a symmetric server side be useful as well?

File: lwt/examples/tls_over_tls.ml [hunk 3]
```diff
@@ -0,0 +1,103 @@
+open Lwt
```
[the remark this answers is not in the document]
Comment (art-w) [reply lwt/examples/tls_over_tls.ml:1]: Thanks, I didn't realize the example could be simplified to get rid of tickets and redundant bits! I like that this test is self-contained, but I may have gone too far: Should I had some cmdliner options to define the proxy and target server?

File: lwt/examples/tls_over_tls.ml [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+  Tls.Config.lookup = (fun _ -> None) ;
+  ticket_granted = (fun psk epoch -> mypsk := Some (psk, epoch)) ;
+  lifetime = 0l ;
+  timestamp = Ptime_clock.now
+}
+
+let authenticator = null_auth
+let client =
+  Tls.Config.client
+    ~version:(`TLS_1_0, `TLS_1_3)
```
[the remark this answers is not in the document]
Comment (art-w) [reply lwt/examples/tls_over_tls.ml]: Thanks, I removed the optional `~version` and it's now using the default (which is the same as your recommendation!)

**Comment by art-w** [comment]
Thanks for the review!

`tls-eio` already supported this, as the client/server are using two-way `Flow` as input and output. For `tls-async`, I added a commit to allow clients to skip the `Tcp.connect` step, and use existing `Reader`/`Writer` instead.

Considering the code organization, it was easy to also support server over channels instead of file descriptors... Although I don't personally have a use-case for this :) 

(I've already rebased this PR to get rid of Cstructs in anticipation of #497 , so it shouldn't be a burden to update it once that's merged)

**Comment by hannesm** [comment]
This should be rebased on top of main (#497 got merged). I still haven't reviewed in much detail, and am a bit hesitant to merge.

File: lwt/tls_lwt.ml [hunk 5]
```diff
@@ -276,11 +322,7 @@ type oc = Lwt_io.output_channel
 let of_t ?close t =
   let close = match close with
     | Some f -> (fun () -> Unix.safely (f ()))
-    | None   -> (fun () ->
-        (* (partially) avoid double-closes by checking if the fd has already been closed *)
-        match Lwt_unix.state t.Unix.fd with
-        | Lwt_unix.Closed -> Lwt.return_unit
-        | Lwt_unix.Opened | Lwt_unix.Aborted _ -> Unix.(safely (close t)))
+    | None   -> (fun () -> Unix.(safely (close t)))
```
Comment (hannesm) [inline lwt/tls_lwt.ml:325]: sorry I am puzzled why this change is part of this PR? And is the change intentional? If yes, please elaborate.

**Comment by hannesm** [comment]
> This should be rebased on top of main (#497 got merged). I still haven't reviewed in much detail, and am a bit hesitant to merge.

Actually, once rebased and the review comments are cleared, it is fine to merge.

File: lwt/tls_lwt.ml (hunk 5 above)
Comment (hannesm) [reply lwt/tls_lwt.ml:331]: ah, it moved into the Lwt_fd module, and inside of Unix.t we now have a Lwt_fd.t

**Comment by hannesm** [comment]
I rebased, and when CI is happy I plan to merge & release.

**Comment by art-w** [comment]
Thanks a ton! (again sorry for the late reply, I was on break without internet)
